File: Example/OpAdxSdkDemo_OC/OpAdxDemo_OC/Classes/InterstitialViewController.m
//
//  InterstitialViewController.m
//  OpAdxDemo_OC
//
//  Created by Luan Chen on 2025/12/10.
//

#import "InterstitialViewController.h"
#import <OpAdxSdk/OpAdxSdk.h>


@interface InterstitialViewController ()<OpAdxInterstitialAdDelegate>

@property (nonatomic, strong, nullable) OpAdxInterstitialAdBridge *interstitialAd;

@end

@implementation InterstitialViewController

- (BOOL)hasVideo {
    return YES;
}

- (AdFormat)adFormat {
    return AdFormatInterstitial;
}

- (NSString *)adFormatString {
    return @"Interstitial Ad";
}

- (void)loadAd {
    if (!self.placementId) return;

    [self.logView print:@"Loading ..."];
    [OpAdxLogger logAdLoadStartWithAdFormat:@"Interstitial" placementId:self.placementId];

    if (self.interstitialAd != nil) {
        [self destroyAd];
    }

    OpAdxInterstitialAdBridge *interstitialAd = [[OpAdxInterstitialAdBridge alloc] initWithPlacementId:self.placementId auctionType: AdAuctionTypeRegular];
    self.interstitialAd = interstitialAd;
    interstitialAd.delegate = self;
    [interstitialAd loadAd];
}

- (void)showAd {
    if (!self.interstitialAd) return;
    
    if (![self.interstitialAd isAdValid]) {
        [self.logView print:@"Ad is invalidated."];
        [self destroyAd];
        return;
    }
    
    __weak typeof(self) weakSelf = self;
    dispatch_after(dispatch_time(DISPATCH_TIME_NOW, (int64_t)(0.5 * NSEC_PER_SEC)), dispatch_get_main_queue(), ^{
        [weakSelf.interstitialAd showAdFrom:weakSelf];
    });
    
    [self enableDestroyAd];
    [self disableShowAd];
}

- (void)destroyAd {
    [super destroyAd];
    self.interstitialAd = nil;
    [self disableDestroyAd];
    [self disableShowAd];
}

#pragma mark - OpAdxInterstitialAdDelegate

- (void)interstitialAdDidLoad:(OpAdxInterstitialAdBridge *)interstitialAd {
    NSLog(@"[ADX] Interstitial广告加载成功: %@", interstitialAd.placementId);
    [OpAdxLogger logAdLoadSuccessWithAdFormat:@"Interstitial" placementId:interstitialAd.placementId ecpm:0];
    [self.logView print:[NSString stringWithFormat:@"Loaded."]];

    [self enableShowAd];
    [self enableDestroyAd];

}

- (void)interstitialAd:(OpAdxInterstitialAdBridge *)interstitialAd didFailWithError:(OpAdxAdError *)error {
    NSLog(@"[ADX] Interstitial广告加载失败: %@ %@", interstitialAd.placementId, error);
    [OpAdxLogger logAdLoadFailedWithAdFormat:@"Interstitial" placementId:interstitialAd.placementId error:error];
    [self.logView print:error.message];

}

- (void)interstitialAdDidClick:(OpAdxInterstitialAdBridge *)interstitialAd {
    NSLog(@"[ADX] Interstitial广告被点击: %@", interstitialAd.placementId);
    [OpAdxLogger logAdClickWithAdFormat:@"Interstitial" placementId:interstitialAd.placementId];
    [self.logView print:@"onAdDidClick"];

}

- (void)interstitialAdDidClose:(OpAdxInterstitialAdBridge *)interstitialAd {
    NSLog(@"[ADX] Interstitial广告关闭: %@", interstitialAd.placementId);
    [OpAdxLogger logAdClosedWithAdFormat:@"Interstitial" placementId:interstitialAd.placementId];
    [self.logView print:@"onAdDidClose"];
    // 清理
    self.interstitialAd = nil;

}

- (void)interstitialAdDidDisplay:(OpAdxInterstitialAdBridge *)interstitialAd {
    NSLog(@"[ADX] Interstitial广告展示: %@", interstitialAd.placementId);
    [OpAdxLogger logAdImpressionWithAdFormat:@"Interstitial" placementId:interstitialAd.placementId];
    [self.logView print:@"onAdImpression"];

}

@end

File: Example/OpAdxSdkDemo_Swift/OpAdxSdkDemo/InterstitialViewController.swift
//
//  InterstitialViewController.swift
//  OperaAdxSDK
//
//  Created by Luan Chen on 2025/11/4.
//

import UIKit
import OpAdxSdk

class InterstitialViewController: BaseViewController {
    
    var interstitialAd: OpAdxInterstitialAd?
    
    override var hasVideo: Bool {
        return true
    }
    
    override var adFormat: AdFormat {
        return .interstitial
    }
    
    override var adFormatString: String {
        return "Interstitial Ad"
    }
    
    override func loadAd() {
        guard let placementId = placementId else { return }

        logView.print("Loading ...")
        OpAdxLogger.logAdLoadStart(adFormat: "Interstitial", placementId: placementId)

        if interstitialAd != nil {
            self.destroyAd()
        }
        interstitialAd = OpAdxInterstitialAd(placementId: placementId, auctionType: AdAuctionType.regular)

        // 创建监听器实例
        let listener = OpAdxInterstitialAdLoadListenerImp(
            onAdLoaded: { [weak self] ad in
                guard let self = self else { return }

                self.logView.print("Loaded")
                OpAdxLogger.logAdLoadSuccess(adFormat: "Interstitial", placementId: placementId)
                self.enableShowAd()
                self.enableDestroyAd()
            },
            onAdFailedToLoad: { [weak self] error in
                guard let self = self else { return }
                self.logView.print(error.message)
                OpAdxLogger.logAdLoadFailed(adFormat: "Interstitial", placementId: placementId, error: error)
            }
        )

        interstitialAd?.load(placementId: placementId, listener: listener)
    }
    
    override func showAd() {
        guard let interstitialAd = interstitialAd else { return }
        
        if interstitialAd.isAdInvalidated() {
            logView.print("Ad is invalidated.")
            destroyAd()
            return
        }
        
        // 创建交互监听器
        let interactionListener = OpAdxInterstitialAdInteractionListenerImp(
            onAdClicked: { [weak self] in
                guard let self = self else { return }
                self.logView.print("Clicked!")
                OpAdxLogger.logAdClick(adFormat: "Interstitial", placementId: self.placementId ?? "unknown")
            },
            onAdDisplayed: { [weak self] in
                guard let self = self else { return }
                self.logView.print("Displayed!")
                OpAdxLogger.logAdImpression(adFormat: "Interstitial", placementId: self.placementId ?? "unknown")
            },
            onAdDismissed: { [weak self] in
                guard let self = self else { return }
                self.logView.print("Dismissed")
                OpAdxLogger.logAdClosed(adFormat: "Interstitial", placementId: self.placementId ?? "unknown")
            },
            onAdFailedToShow: { [weak self] error in
                guard let self = self else { return }
                self.logView.print(error.message)
                OpAdxLogger.logError("Interstitial ad failed to show: \(error.message)")
            }
        )
        
        interstitialAd.show(on: self, listener: interactionListener)
        
        enableDestroyAd()
        disableShowAd()
    }
    
    override func destroyAd() {
        super.destroyAd()
        interstitialAd?.destroy()
        interstitialAd = nil
        disableDestroyAd()
        disableShowAd()
    }
    deinit {
        destroyAd()
    }
}
